R/feature_cvpat.R

```diff
@@ -0,0 +1,229 @@
+    for (iter in index) {
+       results <- rbind(results,c(iter,bootstrap_cvpat(loss_one[,iter],
+                                       loss_two[,iter],
+                                       testtype = testtype,
+                                       BootSamp = BootSamp)))
+    }
+return(results)
+}
+
+lv_loss <- function(construct, model, error) {
+  if(class(error) == "data.frame" | class(error) == "matrix") {
+    loss <- rowMeans(error[,seminr:::items_of_construct(construct = construct,
+                                                        model = model),drop = F]^2)
+  }
+  if (class(error) == "numeric") {
+    loss <- error[,seminr:::items_of_construct(construct = construct,
+                                               model = model),drop = F]^2
+  }
+  return(loss)
+}
+
+overall_loss <- function(error) {
+  if(class(error)[1] == "data.frame" | class(error)[1] == "matrix") {
+    loss <- rowMeans(error)
+  }
+  if (class(error)[1] == "numeric") {
+    loss <- error
+  }
+  return(loss)
+}
+
+bootstrap_cvpat <- function(LossM1,
+                            LossM2,
+                            testtype = "two.sided",
+                            BootSamp = 2000) {
+
+  N <- length(LossM1)
+
+
+
+  OrgTtest<-t.test(LossM2,LossM1,alternative = testtype, paired=TRUE)$statistic
+
+
+  # Originial average difference in losses
+  OrgDbar<-mean(LossM2-LossM1)
+  # Differences in loss functions under the null
+  D_0<-LossM2-LossM1-OrgDbar
+  # Differences in loss functions
+  D<-LossM2-LossM1
+  #Allocating memory to bootrap
+  BootSample <- matrix(0,ncol=2,nrow=length(D))
+  BootDbar <- rep(0,BootSamp)
+  m_losses<-cbind(LossM1,LossM2)
+  tStat <- rep(0,BootSamp)
+  for (b in 1:BootSamp) {
+    BootSample <- m_losses[sample((1:length(D)), length(D), replace=TRUE),]
+    tStat[b]<-t.test(BootSample[,2],BootSample[,1],mu=mean(D),alternative=testtype, paired=TRUE)$statistic
+    BootDbar[b] <- mean(sample(D_0, length(D_0), replace=TRUE))
+  }
+  SorttStat<-sort(tStat, decreasing = FALSE)
+  SortBootDbar<-sort(BootDbar, decreasing = FALSE)
+  # Bootstrap variance on Dbar for t-test
+  std<-sqrt(var(BootDbar))
+  tstat_boot_Var<-OrgDbar/std
+  # Calculating p-values
+  if (testtype=="two.sided") {
+    p.value_perc_Ttest<-(sum(SorttStat>abs(OrgTtest))+sum(SorttStat<=(-abs(OrgTtest))))/BootSamp
+    p.value_perc_D<-(sum(SortBootDbar>abs(OrgDbar))+sum(SortBootDbar<=(-abs(OrgDbar))))/BootSamp
+    p.value_var_ttest<-2*pt(-abs(tstat_boot_Var),(N-1), lower.tail = TRUE)
+  }
+  if (testtype=="greater") {
+    p.value_perc_Ttest<-1-(head(which(SorttStat>OrgTtest),1)-1)/(BootSamp+1)
+    p.value_perc_D<-1-(head(which(SortBootDbar>OrgDbar),1)-1)/(BootSamp+1)
+    p.value_var_ttest<-pt(tstat_boot_Var,(N-1), lower.tail = FALSE)
+    if (length(which(SorttStat>OrgTtest))==0){
+      p.value_perc_Ttest=0
+      p.value_perc_D=0
+    }
+  }
+  # Load outputs 1
+  Results1 <- c("Std. T value"=OrgTtest, "Std. P value"=p.value_perc_Ttest,
```

Comment:
consider giving this return object a class of some kind so that we can write a customized print() function for it if needed.